src/main/scala/scorex/core/ModifiersCache.scala

```diff
@@ -0,0 +1,136 @@
+
+  protected val cache = mutable.Map[K, V]()
+
+  protected val rememberedKeys = mutable.HashSet[K]()
+
+  /**
+    * Defines a best (and application-specific) candidate to be applied.
+    * @param history - an interface to history which could be needed to define a candiate
+    * @return - candidate if it is found
+    */
+  def findCandidateKey(history: H): Option[K]
+
+  protected def onPut(key: K): Unit = {}
+  protected def onRemove(key: K, rememberKey: Boolean): Unit = {}
+
+  /**
+    * A cache element replacement strategy method, which defines a key to remove from cache when it is overfull
+    */
+  protected def keyToRemove(): K
+
+
+  def contains(key: K): Boolean = cache.contains(key) || rememberedKeys.contains(key)
+
+  def put(key: K, value: V): Unit = synchronized {
+    onPut(key)
+    cache.put(key, value)
+    currentCacheSize = currentCacheSize + 1
+
+    if (currentCacheSize > maxSize) remove(keyToRemove())
+  }
+
+  /**
+    * Remove an element from the cache.
+    * @param key - modifier's key
+    * @param rememberKey - whether to remember the key as belonging to cache. E.g. invalid modifiers are
+    *                    to be remembered (for not to be requested from the network again).
+    * @return
+    */
+  def remove(key: K, rememberKey: Boolean = false): Option[V] = synchronized {
+    onRemove(key, rememberKey)
+    currentCacheSize = currentCacheSize - 1
+    if (rememberKey) rememberedKeys += key
+    cache.remove(key)
+  }
+
+  def popCandidate(history: H): Option[V] = synchronized {
+    findCandidateKey(history).flatMap(k => remove(k))
+  }
+}
+
+trait LRUCache[PMOD <: PersistentNodeViewModifier] extends ModifiersCache[PMOD] {
+
+  type Counter = Long
+
+  private var cnt = 0: Counter
+
+  private val keyIndex = mutable.Map[K, Counter]()
+
+  override protected def onPut(key: K): Unit = {
+    keyIndex.put(key, cnt)
+    cnt = cnt + 1
+  }
+
+  override protected def onRemove(key: K, rememberKey: Boolean): Unit = {
+    keyIndex.remove(key)
+  }
+
+  def keyToRemove(): K = {
+    val (key, _) = keyIndex.minBy(_._2)  //todo: can we do faster search?
+    keyIndex.remove(key)
+    key
+  }
+}
+
+class DefaultModifiersCache[PMOD <: PersistentNodeViewModifier, HR <: HistoryReader[PMOD, _]]
+  (override val maxSize: Int) extends ModifiersCache[PMOD] with LRUCache[PMOD] with ScorexLogging {
+
+  override type H = HR
+
+  /**
+    * Default implementation is just about to scan. Not efficient at all and should be probably rewritten in a
+    * concrete application.
+    *
+    * @param history - an interface to history which could be needed to define a candiate
+    * @return - candidate if it is found
+    */
+  @SuppressWarnings(Array("org.wartremover.warts.IsInstanceOf"))
+  override def findCandidateKey(history: H): Option[K] = {
+
+    cache.find { case (k, v) =>
+      history.applicableTry(v) match {
+        case Failure(e) if e.isInstanceOf[RecoverableModifierError] =>
```

Comment:
You can do `Failure(e:RecoverableModifierError)`.

Then you can remove the `SuppressWarnings`  a few lines above.